support/mesh/mesh_channel_core/src/mpsc.rs

```diff
@@ -0,0 +1,971 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+//! Implementation of a multi-producer, single-consumer (MPSC) channel that can
+//! be used to communicate between mesh nodes.
+
+// UNSAFETY: Needed to erase types to avoid monomorphization overhead.
+#![expect(unsafe_code)]
+
+use crate::deque::ElementVtable;
+use crate::deque::ErasedVecDeque;
+use crate::error::ChannelError;
+use crate::error::RecvError;
+use crate::error::TryRecvError;
+use core::fmt::Debug;
+use core::future::Future;
+use core::marker::PhantomData;
+use core::mem::ManuallyDrop;
+use core::mem::MaybeUninit;
+use core::task::Context;
+use core::task::Poll;
+use core::task::Waker;
+use mesh_node::local_node::HandleMessageError;
+use mesh_node::local_node::HandlePortEvent;
+use mesh_node::local_node::Port;
+use mesh_node::local_node::PortField;
+use mesh_node::local_node::PortWithHandler;
+use mesh_node::message::MeshField;
+use mesh_node::message::Message;
+use mesh_protobuf::DefaultEncoding;
+use mesh_protobuf::Protobuf;
+use parking_lot::Mutex;
+use parking_lot::MutexGuard;
+use std::sync::Arc;
+use std::sync::OnceLock;
+
+/// Creates a new channel for sending messages of type `T`, returning the sender
+/// and receiver ends.
+pub fn channel<T>() -> (Sender<T>, Receiver<T>) {
+    let (sender, receiver) = channel_core(const { &ElementVtable::new::<T>() });
+    (Sender(sender, PhantomData), Receiver(receiver, PhantomData))
+}
+
+fn channel_core(vtable: &'static ElementVtable) -> (SenderCore, ReceiverCore) {
```

Comment:
style: my pref is to put these sorts of monomorphized helper functions into the method that uses them directly, thereby sidesteping the need to leak them into the module scope.

i.e: `fn channel<T>() { fn channel_inner() { ... }; channel_inner() }`